llama/generation.py

```diff
@@ -94,124 +108,175 @@ def build(
+        if checkpoint:
+            model.load_state_dict(checkpoint, strict=False)
+        model = model.to(device)
         print(f"Loaded in {time.time() - start_time:.2f} seconds")
 
-        return Llama(model, tokenizer)
+        return Llama(model, tokenizer, device, dynamo)
 
-    def __init__(self, model: Transformer, tokenizer: Tokenizer):
+    def __init__(self, model: Transformer, tokenizer: Tokenizer, device: torch.device, dynamo: bool = True):
         self.model = model
         self.tokenizer = tokenizer
-        self._generate_one_token_fn = self._generate_one_token
-        self._generate_one_token_fn = torch.compile(self._generate_one_token_fn,
-                                                    backend="torchxla_trace_once", fullgraph=True)
-
+        self.device = device
 
-    def _generate_one_token(self, tokens, input_tokens, input_text_mask, cur_pos_tensor, 
-                            input_pos_tensor, output_pos_tensor, cache_kvs, temperature, top_p, logprobs, prev_pos):
-        logits, cache_kvs = self.model(input_tokens, input_pos_tensor, output_pos_tensor, cache_kvs)
+        self._generate_one_token_fn = self._generate_one_token
+        if dynamo:
+            if USE_CUDA:
+                # Inductor errors out when compiles _generate_one_token_fn.
+                # TODO(alanwaketan): figure out why.
+                self.model = torch.compile(self.model, fullgraph=True)
+            else:
+                self._generate_one_token_fn = torch.compile(
+                    self._generate_one_token_fn,
+                    backend="torchxla_trace_once",
+                    fullgraph=True)
+            
+    def _generate_one_token(self, tokens, input_tokens, input_text_mask,
+                            cur_pos_tensor, input_pos_tensor,
+                            output_pos_tensor, temperature_tensor,
+                            top_p_tensor, with_temp, logprobs, token_logprobs, eos_reached, pad_id):
         if logprobs:
-            token_logprobs[:, prev_pos + 1 : cur_pos + 1] = -F.cross_entropy(
-                input=logits.transpose(1, 2),
-                target=tokens[:, prev_pos + 1 : cur_pos + 1],
-                reduction="none",
-                ignore_index=pad_id,
-            )
-
-        if temperature > 0:
-            probs = torch.softmax(logits[:, -1] / temperature, dim=-1)
-            next_token = sample_top_p(probs, top_p)
+            full_logits = self.model(input_tokens, input_pos_tensor, None)
+            logits = full_logits.index_select(1, output_pos_tensor - input_pos_tensor[0]).squeeze(dim=1)
         else:
-            next_token = torch.argmax(logits[:, -1], dim=-1)
+            logits = self.model(input_tokens, input_pos_tensor, output_pos_tensor)
+        if with_temp:
+            probs = torch.softmax(logits / temperature_tensor, dim=-1)
+            next_token = sample_top_p(probs, top_p_tensor)
+        else:
+            next_token = torch.argmax(logits, dim=-1)
```

Comment:
ditto

Reply:
I've optimized the `model.py` code here: https://github.com/pytorch-tpu/llama/blob/42fb3340b52fcbdfde0a36823d2649b567817356/llama/model.py#L410-L412